app/api/entities/entitySchema.js

```diff
@@ -0,0 +1,304 @@
+};
+
+const validateNumericProperty = (property, value) => {
+  if (property.type === templateTypes.numeric) {
+    return isNumber(value);
+  }
+  return true;
+};
+
+const validateDateProperty = (property, value) => {
+  if (property.type === templateTypes.date) {
+    return isNumber(value) && value >= 0;
+  }
+  return true;
+};
+
+const validateMultiDateProperty = (property, value) => {
+  if (property.type === templateTypes.multidate) {
+    return Array.isArray(value) && value.every(isNumber);
+  }
+  return true;
+};
+
+const validateDateRangeProperty = (property, value) => {
+  if (property.type === templateTypes.daterange) {
+    return isValidDateRange(value);
+  }
+  return true;
+};
+
+const validateMultiDateRangeProperty = (property, value) => {
+  if (property.type === templateTypes.multidaterange) {
+    return value.every(isValidDateRange);
+  }
+  return true;
+};
+
+const validateSelectProperty = (property, value) => {
+  if (property.type === templateTypes.select) {
+    return isValidSelect(value);
+  }
+  return true;
+};
+
+const validateGeolocationProperty = (property, value) => {
+  if (property.type === templateTypes.geolocation) {
+    return Array.isArray(value) && value.every(isValidGeolocation);
+  }
+  return true;
+};
+
+const validateMultiSelectProperty = (property, value) => {
+  if ([templateTypes.multiselect, templateTypes.relationship].includes(property.type)) {
+    return Array.isArray(value) && value.every(isValidSelect);
+  }
+  return true;
+};
+
+const validateLinkProperty = (property, value) => {
+  if (property.type === templateTypes.link) {
+    return isString(value.label) && value.label && isString(value.url) && value.url;
+  }
+  return true;
+};
+
+
+const validateMetadataField = (property, entity) => {
+  const value = entity.metadata[property.name];
+  if (!validateRequiredProperty(property, value)) {
+    return false;
+  }
+  if (isUndefined(value) || isNull(value)) {
+    return true;
+  }
+  const propertyValidators = [
+    validateDateProperty,
+    validateMultiDateProperty,
+    validateDateRangeProperty,
+    validateMultiDateRangeProperty,
+    validateTextProperty,
+    validateNumericProperty,
+    validateSelectProperty,
+    validateMultiSelectProperty,
+    validateLinkProperty,
+    validateGeolocationProperty
+  ];
+  return propertyValidators.every(validate => validate(property, value));
+};
+
+ajv.addKeyword('metadataMatchesTemplateProperties', {
+  async: true,
+  errors: false,
+  type: 'object',
+  async validate(schema, entity) {
+    const [template] = await templatesModel.get({ _id: entity.template });
+    if (!template) {
+      return false;
+    }
+
+    return template.properties.every((property) => validateMetadataField(property, entity));
+  }
+});
+
+const schema = {
+  $schema: 'http://json-schema.org/schema#',
+  $async: true,
+  type: 'object',
+  required: ['title'],
+  metadataMatchesTemplateProperties: true,
+  definitions: {
+    dateRange: {
```

Comment:
I would prefer to pull out definitions into separate schema objects so they can be used in other schemas as well

const dateRangeSchema = { type: ...}
...
const schema = { ...
	definitions: { dateRangeSchema }
...